File: libdriveless/tests/src/test_waypoint.cpp
#include <gtest/gtest.h>
#include "../../include/waypoint.h"
#include "test_utils.h"
#include <cmath>

TEST(StateWaypoint, TestSetGet)
{
    Waypoint p(16, 10, angle::deg(60));
    ASSERT_EQ(16, p.x());
    ASSERT_EQ(10, p.z());
    ASSERT_DEQ(60, p.heading().deg());
}


TEST(StateWaypoint, TestOperators)
{
    Waypoint p1(16, 10, angle::deg(60));
    Waypoint p2(16, 10, angle::deg(60));
    Waypoint p3(17, 10, angle::deg(60));
    Waypoint p4(16, 11, angle::deg(60));
    Waypoint p5(16, 11, angle::deg(61));

    ASSERT_TRUE(p1 == p2);
    ASSERT_FALSE(p1 != p2);
    ASSERT_TRUE(p1 != p3);
    ASSERT_TRUE(p1 != p4);
    ASSERT_TRUE(p1 != p5);
    
}


TEST(StateWaypoint, TestComputeDistance)
{
    Waypoint p1(0, 0, angle::deg(0));
    Waypoint p2(10, 10, angle::deg(0));
    Waypoint p3(-10, -10, angle::deg(0));
    ASSERT_DEQ(Waypoint::distanceBetween(p1, p2), 10 * sqrt(2));
    ASSERT_DEQ(Waypoint::distanceBetween(p2, p2), 0);
    ASSERT_DEQ(Waypoint::distanceBetween(p3, p2), 20 * sqrt(2)); 
}

TEST(StateWaypoint, TestComputeHeading)
{
    Waypoint p1(0, 0, angle::deg(0));
    Waypoint p2(1, 1, angle::deg(0));
    Waypoint p3(2, 0, angle::deg(0));
    Waypoint p4(2, 1, angle::deg(0));
    Waypoint p5(0, 1, angle::deg(0));
    Waypoint p6(2, 2, angle::deg(0));
    Waypoint p7(0, 2, angle::deg(0));
    Waypoint p8(0, 4, angle::deg(0));


    ASSERT_TRUE(Waypoint::computeHeading(p1, p1) == angle::deg(0));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p1) == angle::deg(-45));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p3) == angle::deg(45));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p4) == angle::deg(90));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p5) == angle::deg(-90));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p6) == angle::deg(135));
    ASSERT_TRUE(Waypoint::computeHeading(p2, p7) == angle::deg(225));
    ASSERT_TRUE(Waypoint::computeHeading(p1, p8) == angle::deg(180));
}

File: libdriveless/tests/python/test_waypoint.py
import sys, time
sys.path.append("../../")
sys.path.append("../")
import unittest, math
from pydriveless import angle
from pydriveless import Waypoint

class TestWaypoint(unittest.TestCase):
    
    def test_waypoint_encode_decode(self):
        p1 = Waypoint(10, 11, angle.new_deg(20.34))
        s = str(p1)
        p2 = Waypoint.from_str(s)
        
        self.assertEqual(p1.x, p2.x)
        self.assertEqual(p1.z, p2.z)
        self.assertTrue(p1.heading == p2.heading)

    def test_waypoint_distance_between(self):
        p1 = Waypoint(0, 0)
        p2 = Waypoint(10, 10)
        self.assertAlmostEqual(Waypoint.distance_between(p1, p2), 10 * math.sqrt(2), 4)
        p1 = Waypoint(10, 10)
        p2 = Waypoint(10, 10)
        self.assertAlmostEqual(Waypoint.distance_between(p1, p2), 0, 4)
        p1 = Waypoint(-10, -10)
        p2 = Waypoint(10, 10)
        self.assertAlmostEqual(Waypoint.distance_between(p1, p2), 20 * math.sqrt(2), 4)

    def test_waypoint_compute_heading(self):
        p1 = Waypoint(0, 0)
        p2 = Waypoint(0, 0)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 0, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(0, 0)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), -45, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(2, 0)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 45, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(2, 1)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 90, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(0, 1)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), -90, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(2, 2)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 135, 4)
        p1 = Waypoint(1, 1)
        p2 = Waypoint(0, 2)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 225, 4)
        p1 = Waypoint(0, 0)
        p2 = Waypoint(0, 4)
        self.assertAlmostEqual(Waypoint.compute_heading(p1, p2).deg(), 180, 4)


    def test_clone (self) -> None:
        p1 = Waypoint(x=1, z=2, heading=angle.new_deg(3))
        p2 = p1.clone()
        self.assertEqual(p1.x, p2.x)
        self.assertEqual(p1.z, p2.z)
        self.assertEqual(p1.heading, p2.heading)
        self.assertEqual(p1, p2)
        self.assertFalse(p1 is p2)

if __name__ == "__main__":
    unittest.main()
